src/AaveV2CollectorContractConsolidation.sol

```diff
@@ -0,0 +1,144 @@
+        assets[ABUSD] = Asset(364085e16, 75, 0x833D8Eb16D306ed1FbB5D7A2E019e106B960965A, false);
+        assets[BUSD] = Asset(3399064e14, 75, 0x833D8Eb16D306ed1FbB5D7A2E019e106B960965A, false);
+        assets[ZRX] = Asset(10719e16, 300, 0x2885d15b8Af22648b98B122b22FDF4D2a56c6023, false);
+        assets[AZRX] = Asset(877140e16, 300, 0x2885d15b8Af22648b98B122b22FDF4D2a56c6023, false);
+        assets[AUST] = Asset(89239797e4, 200, address(0), false); // NO FEED
+        assets[ARENFIL] = Asset(41067e16, 300, address(0), false); // NO FEED
+    }
+
+    /// @notice Swaps USDC for specified token
+    /// @param _token the address of the token to swap USDC for
+    /// @param _amountOut the amount of token wanted
+    /// @dev User has to approve USDC transfer prior to calling swap 
+    function swap(address _token, uint256 _amountOut) external {
+        if (_amountOut == 0) revert OnlyNonZeroAmount();
+
+        uint256 amountIn = getAmountIn(_token, _amountOut, ERC20(_token).decimals());
+
+        uint256 quantity = assets[_token].quantity;
+        uint256 sendAmount = _amountOut == type(uint256).max ? quantity : _amountOut;
+
+        unchecked {
+            assets[_token].quantity = quantity - sendAmount;
+        }
+
+        USDC.transferFrom(msg.sender, AaveV2Ethereum.COLLECTOR, amountIn);
+        ERC20(_token).safeTransferFrom(AaveV2Ethereum.COLLECTOR, msg.sender, sendAmount);
+        emit Swap(address(USDC), _token, amountIn, sendAmount);
+    }
+
+    /// @notice Returns amount of USDC to be spent to swap for token
+    /// @param _token the address of the token to swap
+    /// @param _amountOut the amount of token wanted
+    /// @param _decimals decimals of ERC20 token
+    /// return amountInWithDiscount the amount of USDC used minus premium incentive
+    /// @dev User check this function before calling swap() to see the amount of USDC required
+    function getAmountIn(
+        address _token,
+        uint256 _amountOut,
+        uint256 _decimals
+    ) public view returns (uint256 amountIn) {
+        Asset memory asset = assets[_token];
+        if (asset.oracle == address(0)) revert UnsupportedToken();
+
+        if (_amountOut == type(uint256).max) {
+            _amountOut = asset.quantity;
+        } else if (_amountOut > asset.quantity) {
+            revert NotEnoughTokens();
+        }
+
+        (uint256 oraclePrice, uint8 oracleDecimals) = getOraclePrice(asset.oracle);
+        unchecked {
+            uint256 exponent = _decimals + oracleDecimals - USDC.decimals();
+
+            if (asset.ethFeedOnly) {
+                (uint256 ethUsdPrice, uint8 ethUsdDecimals) = getOraclePrice(ETH_USD_FEED);
+                oraclePrice *= ethUsdPrice;
+                exponent += ethUsdDecimals;
+            }
+
+            // Basis points arbitrage incentive
+            amountIn = ((_amountOut * oraclePrice / 10**exponent) // Amount in before discount 
+                * (10000 - asset.premium)) / 10000;
```

Comment:
Math here seems to be wrong. If you're given a premium on that asset, then the discount on USD will be `* 10000 / (10000 + asset.premium)`